### PR TITLE
Add support for `guild scheduled events`

### PR DESCRIPTION
**Pull request by Middledot**
## Summary
This pull request is supposed to implement `guild scheduled events` from issue [#210](https://github.com/Pycord-Development/pycord/issues/210) ([discord-api-docs#3586](https://github.com/discord/discord-api-docs/pull/3586))

## Checklist

- [x] HTTP methods
- [x] `MANAGE_EVENTS` permission (already added)
- [x] Gateway
    - [x] state.parse_x for events
    - [x] Scheduled Events Intent
- [x] Scheduled Event Objects/Attrs
    - [x] Cache implementation
    - [x] ScheduledEvent
        - [x] event.edit()
        - [x] event.delete()
        - [x] event.users() (should be similar to channel.history())
    - [x] guild.create_scheduled_event()
    - [x] guild.fetch_scheduled_events()
    - [x] guild.fetch_event()
- [x] Enums
    - [x] Audit log enums (already added)
    - [x] Scheduled event status enums
- [x] Audit log
    - [x] Converter for scheduled events
    - [x] Transform things
    - [x] Converting Location to a Location Object
- [x] Invite related things
    - [x] event_id added to http method
    - [x] event_id added to invite fetch method
    - [x] event fields in create invite
- [x] Documentation
    - [x] guild.x and event.x methods
    - [x] Events (gateway)
    - [x] enums
- [x] Finalize Typehints
    - [x] Invites, etc.
    - [x] Request response typehints
    - [x] scheduled_events.py typehints

## Notes

* The guild events implementation is going to implemented the same way as members and invites (a mix ig)
* This implementation should be tested before merging as I have no idea if this actually works

## Resources
https://github.com/discord/discord-api-docs/pull/4107
https://github.com/discord/discord-api-docs/pull/4113
https://github.com/discord/discord-api-docs/pull/4139
https://github.com/discord/discord-api-docs/pull/4150
https://github.com/discord/discord-api-docs/pull/4152
https://github.com/discord/discord-api-docs/commit/37a25dd805cdb9ba86002149faeb451e5d426f95

https://discord.com/developers/docs/resources/guild-scheduled-event
https://discord.com/developers/docs/topics/gateway#guild-scheduled-event-create
https://discord.com/developers/docs/resources/invite#get-invite

### CODE REVIEW

**Comment by CodeWithSwastik** [comment]
I like `MANAGE EVENTS` is already a permission.

**Comment by Middledot** [comment]
Yeah i ticked it and mentioned that

File: discord/enums.py [hunk 1]
```diff
@@ -625,13 +626,29 @@ def from_datatype(cls, datatype):
...
     betrayal = 773336526917861400
     fishing  = 814288819477020702
     chess    = 832012774040141894
 
+
+class GuildEventStatus(Enum):
+    scheduled = 1
+    active = 2
+    completed = 3
+    canceled = 4
```
Comment (nerdguyahmad) [inline discord/enums.py:642]:
```suggestion
    cancelled = 4
```

File: discord/enums.py [hunk 2]
```diff
@@ -639,7 +639,7 @@ class GuildEventStatus(Enum):
     scheduled = 1
     active = 2
     completed = 3
-    canceled = 4
+    cancelled = 4
```
Comment (nerdguyahmad) [inline discord/enums.py:642]: > I think that’s what it is

Interesting, I didn't know "canceled"  was also a word (just googled it, and it is) but I think aliasing is always an option, like `discord.Color` and `discord.Colour` :grin:

Comment (Middledot) [reply discord/enums.py:642]: yeah probably

**Comment by Lulalaby** [comment]
https://discord.com/developers/docs/resources/guild-scheduled-event
https://discord.com/developers/docs/topics/gateway#guild-scheduled-event-create
https://discord.com/developers/docs/resources/invite#get-invite

**Comment by dtrm-thil** [comment]
@Lulalaby Add cover foto because discord may code it to the api anytime.

**Comment by Lulalaby** [comment]
> 
> 
> @Lulalaby Add cover foto because discord may code it to the api anytime.

Stop stalking..

**Comment by Lulalaby** [comment]
https://github.com/discord/discord-api-docs/pull/4107

**Comment by Lulalaby** [comment]
https://github.com/discord/discord-api-docs/pull/4113/files

**Comment by Lulalaby** [comment]
https://github.com/discord/discord-api-docs/pull/4152
https://github.com/discord/discord-api-docs/pull/4150
https://github.com/discord/discord-api-docs/pull/4139

**Comment by VincentRPS** [comment]
is this pull stale?

**Comment by Lulalaby** [comment]
Kinda

**Comment by Lulalaby** [comment]
@Middledot good work so far.
Keep it up!

File: discord/commands/errors.py [hunk 3]
```diff
@@ -1,64 +1,64 @@
...
+
+    Attributes
+    -----------
+    original: :exc:`Exception`
+        The original exception that was raised. You can also get this via
+        the ``__cause__`` attribute.
+    """
+    def __init__(self, e: Exception) -> None:
+        self.original: Exception = e
+        super().__init__(f'Application Command raised an exception: {e.__class__.__name__}: {e}')
```
Comment (Lulalaby) [inline discord/commands/errors.py:64]: What happened here?
This is not in scope of scheduled events.

Comment (VincentRPS) [reply discord/commands/errors.py:64]: This seems more like a github error than anything

Comment (Middledot) [reply discord/commands/errors.py:64]: This happened when I started using git to make commits. I used `git push -f` after `git pull` wasn't working and this happened. I checked and there are no actual changes, so yeah, probably like a github error.

Comment (Lulalaby) [reply discord/commands/errors.py:64]: git being git

**Review by Lulalaby — APPROVED** [review]

**Comment by Middledot** [comment]
I think there should be more than 2 reviews before merging

**Review by Dorukyum — APPROVED** [review]
Looks good to me

**Review by Lulalaby — APPROVED** [review]
@BobDotCom please do an additional review

**Comment by Lulalaby** [comment]
And for gods sake, please squash

**Review by ChickenDevs — APPROVED** [review]
Also looks good to me, still waiting on @BobDotCom

**Comment by honzajavorek** [comment]
I just wanted to say that I installed pycord with support for scheduled events as soon as you merged this PR and it works and I love you all.

**Comment by Middledot** [comment]
Glad to hear it works for you!

**Pull request merged**